File: my-app/server/train_mnist.py
#!/usr/bin/env python3
"""
Train MNIST model for digit recognition
Also downloads and prepares an extended dataset with math symbols
"""

import os
import sys
import gzip
import struct
import urllib.request
import numpy as np

# Check for TensorFlow
try:
    import tensorflow as tf
    from tensorflow import keras
    print(f"TensorFlow version: {tf.__version__}")
except ImportError:
    print("TensorFlow not found. Installing...")
    os.system("pip3 install tensorflow")
    import tensorflow as tf
    from tensorflow import keras

MNIST_URLS = {
    'train_images': 'http://yann.lecun.com/exdb/mnist/train-images-idx3-ubyte.gz',
    'train_labels': 'http://yann.lecun.com/exdb/mnist/train-labels-idx1-ubyte.gz',
    'test_images': 'http://yann.lecun.com/exdb/mnist/t10k-images-idx3-ubyte.gz',
    'test_labels': 'http://yann.lecun.com/exdb/mnist/t10k-labels-idx1-ubyte.gz',
}

DATA_DIR = os.path.join(os.path.dirname(__file__), 'data')
MODEL_DIR = os.path.join(os.path.dirname(__file__), '..', 'public', 'models', 'mnist')


def download_mnist():
    """Download MNIST dataset"""
    os.makedirs(DATA_DIR, exist_ok=True)
    
    for name, url in MNIST_URLS.items():
        filepath = os.path.join(DATA_DIR, f'{name}.gz')
        if not os.path.exists(filepath):
            print(f"Downloading {name}...")
            urllib.request.urlretrieve(url, filepath)
            print(f"  Saved to {filepath}")
        else:
            print(f"{name} already exists")


def load_mnist_images(filepath):
    """Load MNIST images from gzipped file"""
    with gzip.open(filepath, 'rb') as f:
        magic, num, rows, cols = struct.unpack('>IIII', f.read(16))
        images = np.frombuffer(f.read(), dtype=np.uint8)
        images = images.reshape(num, rows, cols, 1)
    return images


def load_mnist_labels(filepath):
    """Load MNIST labels from gzipped file"""
    with gzip.open(filepath, 'rb') as f:
        magic, num = struct.unpack('>II', f.read(8))
        labels = np.frombuffer(f.read(), dtype=np.uint8)
    return labels


def create_model():
    """Create a CNN model for digit recognition"""
    model = keras.Sequential([
        # First conv block
        keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        keras.layers.BatchNormalization(),
        keras.layers.Conv2D(32, (3, 3), activation='relu'),
        keras.layers.BatchNormalization(),
        keras.layers.MaxPooling2D((2, 2)),
        keras.layers.Dropout(0.25),
        
        # Second conv block
        keras.layers.Conv2D(64, (3, 3), activation='relu'),
        keras.layers.BatchNormalization(),
        keras.layers.Conv2D(64, (3, 3), activation='relu'),
        keras.layers.BatchNormalization(),
        keras.layers.MaxPooling2D((2, 2)),
        keras.layers.Dropout(0.25),
        
        # Dense layers
        keras.layers.Flatten(),
        keras.layers.Dense(256, activation='relu'),
        keras.layers.BatchNormalization(),
        keras.layers.Dropout(0.5),
        keras.layers.Dense(10, activation='softmax')
    ])
    
    model.compile(
        optimizer='adam',
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy']
    )
    
    return model


def augment_data(images, labels):
    """Apply data augmentation"""
    from tensorflow.keras.preprocessing.image import ImageDataGenerator
    
    datagen = ImageDataGenerator(
        rotation_range=10,
        width_shift_range=0.1,
        height_shift_range=0.1,
        zoom_range=0.1,
        shear_range=0.1
    )
    
    return datagen


def train():
    """Train the MNIST model"""
    print("\n=== MNIST Model Training ===\n")
    
    # Download data
    download_mnist()
    
    # Load data
    print("\nLoading data...")
    train_images = load_mnist_images(os.path.join(DATA_DIR, 'train_images.gz'))
    train_labels = load_mnist_labels(os.path.join(DATA_DIR, 'train_labels.gz'))
    test_images = load_mnist_images(os.path.join(DATA_DIR, 'test_images.gz'))
    test_labels = load_mnist_labels(os.path.join(DATA_DIR, 'test_labels.gz'))
    
    print(f"Train images: {train_images.shape}")
    print(f"Test images: {test_images.shape}")
    
    # Normalize
    train_images = train_images.astype('float32') / 255.0
    test_images = test_images.astype('float32') / 255.0
    
    # Create model
    print("\nCreating model...")
    model = create_model()
    model.summary()
    
    # Data augmentation
    datagen = augment_data(train_images, train_labels)
    
    # Train
    print("\nTraining...")
    history = model.fit(
        datagen.flow(train_images, train_labels, batch_size=128),
        epochs=10,
        validation_data=(test_images, test_labels),
        callbacks=[
            keras.callbacks.EarlyStopping(patience=3, restore_best_weights=True),
            keras.callbacks.ReduceLROnPlateau(factor=0.5, patience=2)
        ]
    )
    
    # Evaluate
    print("\nEvaluating...")
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy: {test_acc:.4f}")
    
    # Save model
    os.makedirs(MODEL_DIR, exist_ok=True)
    
    # Save as Keras format
    keras_path = os.path.join(MODEL_DIR, 'model.keras')
    model.save(keras_path)
    print(f"\nModel saved to {keras_path}")
    
    # Also save as TensorFlow.js format if tensorflowjs is available
    try:
        import tensorflowjs as tfjs
        tfjs_path = MODEL_DIR
        tfjs.converters.save_keras_model(model, tfjs_path)
        print(f"TensorFlow.js model saved to {tfjs_path}")
    except ImportError:
        print("\nTo export for browser use, install tensorflowjs:")
        print("  pip3 install tensorflowjs")
        print("  tensorflowjs_converter --input_format keras", keras_path, MODEL_DIR)
    
    return model


if __name__ == '__main__':
    train()

File: my-app/scripts/train-mnist.js
/**
 * Train MNIST model using TensorFlow.js Node
 * Run: node scripts/train-mnist.js
 */

async function main() {
  // Dynamic import for ES modules
  const tf = await import('@tensorflow/tfjs-node');
  const fs = await import('fs');
  const path = await import('path');
  const https = await import('https');
  const zlib = await import('zlib');
  
  const MNIST_BASE_URL = 'https://storage.googleapis.com/cvdf-datasets/mnist/';
  
  async function fetchBuffer(url) {
    return new Promise((resolve, reject) => {
      https.get(url, (res) => {
        const chunks = [];
        res.on('data', chunk => chunks.push(chunk));
        res.on('end', () => {
          const buffer = Buffer.concat(chunks);
          // Decompress gzip
          zlib.gunzip(buffer, (err, decompressed) => {
            if (err) reject(err);
            else resolve(decompressed);
          });
        });
        res.on('error', reject);
      });
    });
  }
  
  function parseImages(buffer) {
    // Skip 16-byte header
    const pixels = new Float32Array((buffer.length - 16));
    for (let i = 16; i < buffer.length; i++) {
      pixels[i - 16] = buffer[i] / 255.0;
    }
    const numImages = (buffer.length - 16) / (28 * 28);
    return tf.tensor4d(pixels, [numImages, 28, 28, 1]);
  }
  
  function parseLabels(buffer) {
    // Skip 8-byte header
    const labels = new Uint8Array(buffer.length - 8);
    for (let i = 8; i < buffer.length; i++) {
      labels[i - 8] = buffer[i];
    }
    return tf.oneHot(tf.tensor1d(labels, 'int32'), 10);
  }
  
  console.log('Downloading MNIST training images...');
  const trainImagesBuffer = await fetchBuffer(MNIST_BASE_URL + 'train-images-idx3-ubyte.gz');
  console.log('Downloading MNIST training labels...');
  const trainLabelsBuffer = await fetchBuffer(MNIST_BASE_URL + 'train-labels-idx1-ubyte.gz');
  console.log('Downloading MNIST test images...');
  const testImagesBuffer = await fetchBuffer(MNIST_BASE_URL + 't10k-images-idx3-ubyte.gz');
  console.log('Downloading MNIST test labels...');
  const testLabelsBuffer = await fetchBuffer(MNIST_BASE_URL + 't10k-labels-idx1-ubyte.gz');
  
  console.log('Parsing data...');
  const trainImages = parseImages(trainImagesBuffer);
  const trainLabels = parseLabels(trainLabelsBuffer);
  const testImages = parseImages(testImagesBuffer);
  const testLabels = parseLabels(testLabelsBuffer);
  
  console.log('Train images shape:', trainImages.shape);
  console.log('Train labels shape:', trainLabels.shape);
  
  // Create model
  console.log('Creating model...');
  const model = tf.sequential();
  
  model.add(tf.layers.conv2d({
    inputShape: [28, 28, 1],
    kernelSize: 3,
    filters: 32,
    activation: 'relu',
  }));
  model.add(tf.layers.batchNormalization());
  model.add(tf.layers.maxPooling2d({ poolSize: 2 }));
  
  model.add(tf.layers.conv2d({
    kernelSize: 3,
    filters: 64,
    activation: 'relu',
  }));
  model.add(tf.layers.batchNormalization());
  model.add(tf.layers.maxPooling2d({ poolSize: 2 }));
  
  model.add(tf.layers.conv2d({
    kernelSize: 3,
    filters: 64,
    activation: 'relu',
  }));
  model.add(tf.layers.batchNormalization());
  
  model.add(tf.layers.flatten());
  model.add(tf.layers.dropout({ rate: 0.25 }));
  model.add(tf.layers.dense({ units: 128, activation: 'relu' }));
  model.add(tf.layers.dropout({ rate: 0.5 }));
  model.add(tf.layers.dense({ units: 10, activation: 'softmax' }));
  
  model.compile({
    optimizer: tf.train.adam(0.001),
    loss: 'categoricalCrossentropy',
    metrics: ['accuracy'],
  });
  
  model.summary();
  
  // Train
  console.log('Training model (this may take a few minutes)...');
  await model.fit(trainImages, trainLabels, {
    epochs: 5,
    batchSize: 128,
    validationSplit: 0.1,
    callbacks: {
      onEpochEnd: (epoch, logs) => {
        console.log(`Epoch ${epoch + 1}/5: loss=${logs.loss.toFixed(4)}, acc=${logs.acc.toFixed(4)}, val_loss=${logs.val_loss.toFixed(4)}, val_acc=${logs.val_acc.toFixed(4)}`);
      },
    },
  });
  
  // Evaluate
  console.log('Evaluating on test set...');
  const evalResult = model.evaluate(testImages, testLabels);
  console.log(`Test loss: ${evalResult[0].dataSync()[0].toFixed(4)}`);
  console.log(`Test accuracy: ${evalResult[1].dataSync()[0].toFixed(4)}`);
  
  // Save model
  const modelDir = path.join(process.cwd(), 'public', 'models', 'mnist');
  if (!fs.existsSync(modelDir)) {
    fs.mkdirSync(modelDir, { recursive: true });
  }
  
  await model.save(`file://${modelDir}`);
  console.log(`Model saved to ${modelDir}`);
  
  // Cleanup
  trainImages.dispose();
  trainLabels.dispose();
  testImages.dispose();
  testLabels.dispose();
  
  console.log('Done!');
}

main().catch(console.error);
